### DISCUSSION TITLE
not defined

### DISCUSSION THREAD

**Discussion by AndreasSchmidt87** [neuefische/questions#251] [Web] [answered]
Hello coding people,

my problem is, if my snake eats the food then i get this error.

<img width="775" alt="Bildschirmfoto 2022-10-12 um 15 15 19" src="https://user-images.githubusercontent.com/76123735/195352962-f8399b3e-d25f-4801-903a-35d683374bec.png">

I don't know why the snakeboard is given upstairs and the food is shown beforehand.
Here is my code:

![Bildschirmfoto 2022-10-12 um 15 10 51](https://user-images.githubusercontent.com/76123735/195353077-c6760a04-e716-4134-be55-dbad8f61205f.png)

Thanks in advance :)

Here the Link to my Pull Request https://github.com/AndySch666/capstone-project/pull/24

**Comment by ThomasKirst** [comment, marked as answer]
Hi Andi,

you are calling your function `generate_food()` without any parameters in two places. You should pass the `snakeboard` and the `snake`.

```js
function generate_food(snakeboard, snake) {
  food_x = random_food(0, snakeboard.width - 10);
  food_y = random_food(0, snakeboard.height - 10);
  snake.forEach(function has_snake_eaten_food(part) {
    const has_eaten = part.x === food_x && part.y === food_y;
    // You shall pass the parameters 👇
    if (has_eaten) generate_food(snakeboard, snake);
  });
}
```

```js
// The parameter snakeboard needs to be added (instead of event)
function moveSnake(snake, directionX, directionY, snakeboard) {
    const head = { x: snake[0].x + directionX, y: snake[0].y + directionY };
    snake.unshift(head);
    const has_eaten_food = snake[0].x === food_x && snake[0].y === food_y;
    if (has_eaten_food) {
      score += 10;
      document.getElementById("score").innerHTML = score;
      // Pass the params! 👇
      generate_food(snakeboard, snake);
    } else {
      snake.pop();
    }
}
```

```js
// setInterval(function onTick() { …
moveSnake(snake, directionX, directionY, snakeboard); // 👈
```

Please change your function names to reflect the camelCase notation.
Also don´t use `innerHTML` for the score display (`useState()` to the rescue!).

Have fun! 🐍

**Reply by ThomasKirst** [reply]
@AndySch666 Problem solved? ;)

**Reply by AndreasSchmidt87** [reply]
@ThomasKirst the problem with snakeboard is solved, but now i get another problem 
<img width="775" alt="Bildschirmfoto 2022-10-13 um 13 54 31" src="https://user-images.githubusercontent.com/76123735/195590083-a12235e1-53b8-4b5f-aa8c-d08d2c46832b.png">
and here the code:
![Bildschirmfoto 2022-10-13 um 13 55 26](https://user-images.githubusercontent.com/76123735/195590111-acd2b3c0-f5ee-4e07-afe6-7507fbc60f61.png)
thanks in advance
Link to Request https://github.com/AndySch666/capstone-project/pull/24

**Reply by ThomasKirst** [reply]
Your state update does not work properly. Try this instead:

```js
function moveSnake(snake, directionX, directionY, snakeboard, score) {
    const head = { x: snake[0].x + directionX, y: snake[0].y + directionY };
    snake.unshift(head);
    const has_eaten_food = snake[0].x === foodX && snake[0].y === foodY;
    if (has_eaten_food) {
      // 👇 Safely rely on the previous state value => pass a callback function to the state update function  
      setScore((previousScore) => (previousScore += 10));
      // document.getElementById('score').innerHTML = score;
      generate_food(snakeboard, snake, score);
      // 👇 Why return the score? Why do you destructure it? It´s simply a numeric value.
      return { ...score };
    } else {
      snake.pop();
    }
  }
```

**Reply by AndreasSchmidt87** [reply]
@ThomasKirst Thanks now the problems are solved :)
